### PR TITLE
Feature/selector core

### PR DESCRIPTION
**Pull request by MikeSquall**
Related to [this issue](https://github.com/capsulajs/selector-service/issues/6)

### CODE REVIEW

File: src/Selector.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import { ItemsRequest, SelectedItemRequest, SelectItemRequest, SelectorInterface, SetItemsRequest } from './api';
```
Comment (dmitriyslaym) [inline src/Selector.ts]: include imports from external modules firstly

File: src/Selector.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import isMatch from 'lodash/isMatch';
+import isEmpty from 'lodash/isEmpty';
+
+export class Selector<Item extends Key, Key> implements SelectorInterface<Item, Key> {
+  private readonly data$: BehaviorSubject<Item[]>;
+  private readonly selected$: BehaviorSubject<Item>;
+
+  constructor() {
+    this.data$ = new BehaviorSubject<Item[]>([]);
+    this.selected$ = new BehaviorSubject<Item>({} as Item);
```
Comment (dmitriyslaym) [inline src/Selector.ts]: maybe we want to use undefined initially, when nothing has been selected yet? In this case we will have a type BehaviorSubject<Item?>

File: src/Selector.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        return reject(new Error(validationMessages.invalidSetItemsRequest));
+      }
+      this.resetSelected(setItemsRequest.items);
+      this.data$.next(setItemsRequest.items);
+      return resolve();
+    });
+  }
+
+  public items$(itemsRequest: ItemsRequest): Observable<Item[]> {
+    return this.data$;
```
Comment (dmitriyslaym) [inline src/Selector.ts]: You need to use **toObservable()** to remove the redundant features of Subject from the entity that you're returning.

File: src/Selector.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      // Reject case: no data has been set in Selector
+      if (this.data$.getValue().length === 0) {
+        return reject(new Error(errorMessages.noData));
+      }
+      // Reject case: selectItemRequest is not valid
+      if (!isValidSelectRequest(selectItemRequest)) {
+        return reject(new Error(validationMessages.invalidSelectItemRequest));
+      }
+      // Reject case: Item is already selected
+      if (isMatch(this.selected$.getValue() as any, selectItemRequest.key as any)) {
```
Comment (dmitriyslaym) [inline src/Selector.ts]: should it be **this.selected$.getValue() as Item**, **selectItemRequest.key as any as Key** ?

File: src/Selector.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        return reject(new Error(errorMessages.itemAlreadySelected));
+      }
+
+      this.data$
+        .pipe(
+          take(1),
```
Comment (dmitriyslaym) [inline src/Selector.ts]: you also have **first()** operator for it

File: src/Selector.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+          map((items) =>
+            items.find((item: Item) => {
+              return isMatch(item as any, selectItemRequest.key as any);
```
Comment (dmitriyslaym) [inline src/Selector.ts]: also **any** typing

File: src/Selector.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            })
+          )
+        )
+        .subscribe((item: any) => {
```
Comment (dmitriyslaym) [inline src/Selector.ts]: any

File: src/Selector.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            return reject(new Error(errorMessages.itemNotFound));
+          }
+          this.selected$.next(item);
+          return resolve();
+        });
+    });
+  }
+
+  public selectedItem$(selectedItemRequest: SelectedItemRequest): Observable<Item> {
+    return this.selected$;
```
Comment (dmitriyslaym) [inline src/Selector.ts]: use toObservable()

File: src/Selector.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  private resetSelected(items: Item[]): void {
+    const selected = this.selected$.getValue();
+    if (!isEmpty(selected)) {
```
Comment (dmitriyslaym) [inline src/Selector.ts]: if you use undefined, when no item is selected, you will not need this helper

File: src/Selector.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      const shouldKeepSelection = items.some((item) => isMatch(item as any, selected as any));
```
Comment (dmitriyslaym) [inline src/Selector.ts]: typing

File: src/Selector.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+      if (!shouldKeepSelection) {
+        this.selected$.next({} as Item);
```
Comment (dmitriyslaym) [inline src/Selector.ts]: suggest to use undefined as a value

File: src/helpers/messages.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+export const errorMessages = {
+  itemNotFound: 'The item you are looking for was not found',
+  itemAlreadySelected: 'The item is already selected',
```
Comment (dmitriyslaym) [inline src/helpers/messages.ts]: **has already been selected** is better

File: tests/features/service-selector.feature [hunk 13]
```diff
@@ -1,92 +1,63 @@
 
-
-Scenario: Call setItemsStream with stream of items and get items at subscription after one emission
-    Given Selector Service with setItemsStream method
-    And   user calls setItemsStream with a stream X
-    And   stream X emits [A, B, C]
+Scenario: Call setItems with items and get items at subscription after one emission
```
Comment (dmitriyslaym) [inline tests/features/service-selector.feature:2]: Just a reminder - we decided to update TC in a separate PR

File: tests/implementations/selector.test.ts [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+  let selector: SelectorInterface<Character, SimpleCharacterKey>;
+
+  beforeEach(() => {
+    selector = new Selector<Character, SimpleCharacterKey>();
+  });
+
+  it('Call setItems with items and get items at subscription after one emission', async (done) => {
+    expect.assertions(1);
+    await selector.setItems({ items: beatles });
+    selector.items$({}).subscribe((items: any) => {
```
Comment (dmitriyslaym) [inline tests/implementations/selector.test.ts]: please fix all **any** (I will not add more comments about it, but there are some)

Why do you need done here? If you put the amount of expectations - test should wait automatically

File: tests/implementations/selector.test.ts [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    await selector.selectItem({ key: { name: 'John' } });
+    await selector.selectItem({ key: { name: 'Paul' } });
+    await selector.selectItem({ key: { name: 'George' } });
+  });
+
+  it('Call selectItem with currently selected item', async () => {
+    expect.assertions(1);
+    await selector.setItems({ items: beatles });
+    await selector.selectItem({ key: { name: 'Ringo' } });
+    await expect(selector.selectItem({ key: { name: 'Ringo' } })).rejects.toEqual(
```
Comment (dmitriyslaym) [inline tests/implementations/selector.test.ts]: I think it's better to return it

File: tests/mocks/items.ts [hunk 16]
```diff
@@ -0,0 +1,28 @@
+export interface Character {
```
Comment (dmitriyslaym) [inline tests/mocks/items.ts]: do we need this interface only for tests?
Even if yes, I would suggest to put it in a types.ts file

Comment (MikeSquall) [reply tests/mocks/items.ts]: Those interfaces are used only for tests, i don't see the value to put them in the `src/`

File: src/api/Selector.ts [hunk 17]
```diff
@@ -29,7 +29,7 @@ export interface Selector<Item extends Key, Key extends object> {
    * @typeparam SelectedItemRequest
    * @return An Observable sequence that describes updates of the selected item.
    */
-  selectedItem$(selectedItemRequest: SelectedItemRequest): Observable<Item>;
+  selectedItem$(selectedItemRequest: SelectedItemRequest): Observable<Item | undefined>;
```
Comment (dmitriyslaym) [inline src/api/Selector.ts]: I think it makes sense to create a type SelectedItem = Item | undefined
It will be more descriptive and reusable

**Review by dmitriyslaym — APPROVED** [review]

**Pull request merged**